File: tests/r_parity/40_qreg.py
"""StatsPAI quantile regression parity (Python side) -- Module 40.

Generates a heteroskedastic linear DGP and runs sp.qreg at tau=0.5.
The R side uses quantreg::rq; Stata uses qreg.

Tolerance: rel < 1e-3 on coefficients (simplex solver converges to
the same vertex); rel < 5e-2 on SEs (different SE methods).
"""
from __future__ import annotations

import numpy as np
import pandas as pd
import statspai as sp

from _common import PARITY_SEED, ParityRecord, dump_csv, write_results


MODULE = "40_qreg"


def make_data(n: int = 500, seed: int = PARITY_SEED) -> pd.DataFrame:
    rng = np.random.default_rng(seed)
    x1 = rng.normal(0, 1, n)
    x2 = rng.normal(0, 1, n)
    eps = rng.normal(0, 1.0 + 0.3 * x1**2, n)
    y = 1.0 + 0.5 * x1 - 0.3 * x2 + eps
    return pd.DataFrame({"y": y, "x1": x1, "x2": x2})


def main() -> None:
    df = make_data()
    dump_csv(df, MODULE)

    res = sp.qreg(data=df, y="y", x=["x1", "x2"], quantile=0.5)
    # detail is a pandas.DataFrame-like dict with parallel arrays
    coef = res.detail["coefficient"]
    se = res.detail["se"]
    var_ = res.detail["variable"]
    # Build dict by variable name
    dct = {str(v): (float(coef[i]), float(se[i])) for i, v in enumerate(var_)}

    rows: list[ParityRecord] = []
    for name in ["const", "x1", "x2"]:
        if name in dct:
            est, sev = dct[name]
            label = "intercept" if name == "const" else name
            rows.append(ParityRecord(
                MODULE, "py",
                statistic=f"beta_{label}",
                estimate=est, se=sev, n=int(len(df))))

    write_results(MODULE, "py", rows,
                  extra={"quantile": 0.5, "engine": "statsmodels"})


if __name__ == "__main__":
    main()

File: tests/r_parity/40_qreg.R
# StatsPAI quantile regression parity (R side) -- Module 40.
#
# Reads data/40_qreg.csv and runs quantreg::rq with tau=0.5.
# Tolerance: rel < 1e-3 on coefficients; rel < 5e-2 on SEs.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(quantreg)
})

MODULE <- "40_qreg"

df <- read_csv_strict(MODULE)

fit <- quantreg::rq(y ~ x1 + x2, data = df, tau = 0.5)
# nid (Powell sandwich) SEs to match Stata qreg's default kernel
sm  <- summary(fit, se = "nid")

co <- coef(fit)
se <- sm$coefficients[, "Std. Error"]

rows <- list(
  parity_row(MODULE, "beta_intercept",
             estimate = unname(co["(Intercept)"]),
             se = unname(se["(Intercept)"]),
             n = nrow(df)),
  parity_row(MODULE, "beta_x1",
             estimate = unname(co["x1"]),
             se = unname(se["x1"]),
             n = nrow(df)),
  parity_row(MODULE, "beta_x2",
             estimate = unname(co["x2"]),
             se = unname(se["x2"]),
             n = nrow(df))
)

write_results(MODULE, rows,
              extra = list(quantile = 0.5,
                           se_method = "nid_Powell",
                           package = "quantreg::rq"))
